Enable Trimmable Assembly

mwasplund opened the pull request:
With this change I am annotating the core assembly as IsTrimmable. This enables warnings for any known issues that may arise when publishing a final project as trimmed. The only warnings present were the use of runtime reflection code to perform the Json serialization. I converted over to using the Json Source Generator to move this logic from runtime to compile time (a net performance win too).

fixes #390

Commit:
Add IsTrimmable flag and fix json serializer warnings

josephdecock commented on src/OidcClient/Infrastructure/LogSerializer.cs:
```diff
@@ -1,9 +1,11 @@
-
 using System.Text.Json;
 using System.Text.Json.Serialization;
+#if NET5_0_OR_GREATER
```
Why do we need this conditional compilation here, when it wasn't necessary in the other PR for IdentityModel itself?

josephdecock commented on src/OidcClient/OidcClient.csproj:
```diff
@@ -4,14 +4,14 @@
-    <TargetFramework>netstandard2.0</TargetFramework>
-    
+
+    <TargetFrameworks>netstandard2.0;net6.0</TargetFrameworks>
```
I'm pretty hesitant to change TargetFrameworks. I assume some of the json serialization options are missing in netstandard2.0? What are we missing?

Commit:
Pull in System.Text.Json to remove conditionals

mwasplund committed:
Merge branch 'main' into masplund/trimmable

josephdecock committed:
Add console program to test trimmable warnings

josephdecock committed:
Update to latest IdentityModel version

This is needed to make this assembly trimmable

josephdecock committed:
Add trimmable analysis to build

josephdecock committed:
Run tests against .net 6, 7, and 8.

josephdecock committed:
Silly typo

josephdecock committed:
Include trimmable analysis project in sln

josephdecock committed:
Treat trimmable analysis warnings as errors

josephdecock committed:
xmldoc typo

josephdecock committed:
Enable trimming warnings only for net6.0

josephdecock commented on src/OidcClient/Infrastructure/LogSerializer.cs:
```diff
@@ -34,9 +34,11 @@ static LogSerializer()
+            return Enabled ?
+                JsonSerializer.Serialize(logObject, (JsonTypeInfo<T>)SourceGenerationContext.Default.GetTypeInfo(typeof(T))) :
+                "Logging has been disabled";
+		}
```
This is a breaking change because the LogSerializer can now only serialize the types in the source generator context. That works fine for our internal usage of the LogSerializer, but the LogSerializer is a public type. Anyone who is using this helper to serialize anything else will get an NRE when GetTypeInfo returns null.

Options:
- Make the LogSerializer internal. 
- Don't enable trimmable
- Make this method private, but keep the original `Serialize(object logObject)` as a public overload with `[RequiresUnreferencedCode]`

josephdecock committed:
Add log serializer overloads (prevents a breaking change)

Changing the log serializer to support trimming means that we need to add the types that we will serialize into a serialization context. Unfortunately, the serializer is a public type that serializes arbitrary objects today. This is incompatible with trimming. The solution is to mark the existing API as [RequiresExternalCode], which gives trimmed callers of the API a warning, and to add overloads that accept the types that are in the  generation context.
